go/gossip/storage.go

```diff
@@ -0,0 +1,354 @@
+		{&s.selectNumSCTs, selectNumSCTs},
+		{&s.selectNumSTHs, selectNumSTHs},
+		{&s.selectFeedback, selectFeedback},
+		{&s.selectSTH, selectSTH}} {
+		if err := prepareStatement(s.db, p); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Close closes the underlying DB storage.
+func (s *Storage) Close() error {
+	return s.db.Close()
+}
+
+func selectThingID(getID *sql.Stmt, thing interface{}) (int64, error) {
+	rows, err := getID.Query(thing)
+	if err != nil {
+		return -1, err
+	}
+	if !rows.Next() {
+		return -1, fmt.Errorf("couldn't look up ID for %v", thing)
+	}
+	var id int64
+	if err = rows.Scan(&id); err != nil {
+		return -1, err
+	}
+	return id, nil
+}
+
+// insertThingOrSelectID will attempt to execute the insert Statement (under transaction tx), if that fails due to
+// a unique primary key constraint, it will look up that primary key by executing the getID Statement.
+// Returns the ID associated with persistent thing, or an error describing the failure.
+func insertThingOrSelectID(tx *sql.Tx, insert *sql.Stmt, getID *sql.Stmt, thing interface{}) (int64, error) {
+	txInsert := tx.Stmt(insert)
+	txGetID := tx.Stmt(getID)
+	r, err := txInsert.Exec(thing)
+	if err != nil {
+		switch e := err.(type) {
+		case sqlite3.Error:
+			if e.Code == sqlite3.ErrConstraint {
+				return selectThingID(txGetID, thing)
+			}
+		}
+		return -1, err
+	}
+	id, err := r.LastInsertId()
+	if err != nil {
+		return -1, err
+	}
+	return id, nil
+}
+
+func (s *Storage) addChainIfNotExists(tx *sql.Tx, chain []string) (int64, error) {
+	flatChain := strings.Join(chain, "")
+	return insertThingOrSelectID(tx, s.insertChain, s.selectChainID, flatChain)
+}
+
+func (s *Storage) addSCTIfNotExists(tx *sql.Tx, sct string) (int64, error) {
+	return insertThingOrSelectID(tx, s.insertSCT, s.selectSCTID, sct)
+}
+
+func (s *Storage) addSCTFeedbackIfNotExists(tx *sql.Tx, chainID, sctID int64) error {
+	stmt := tx.Stmt(s.insertSCTFeedback)
+	_, err := stmt.Exec(chainID, sctID)
+	if err != nil {
+		switch err.(type) {
+		case sqlite3.Error:
+			// If this is a dupe that's fine, no need to return an error
+			if err.(sqlite3.Error).Code != sqlite3.ErrConstraint {
+				return err
+			}
+		default:
+			return err
+		}
+	}
+	return nil
+}
+
+// AddSCTFeedback stores the passed in feedback object.
+func (s *Storage) AddSCTFeedback(feedback SCTFeedback) (err error) {
+	tx, err := s.db.Begin()
+	if err != nil {
+		return err
+	}
+	// If we return a non-nil error, then rollback the transaction.
+	defer func() {
+		if err != nil {
+			tx.Rollback()
+		}
+	}()
+
+	for _, f := range feedback.Feedback {
+		chainID, err := s.addChainIfNotExists(tx, f.X509Chain)
+		if err != nil {
+			return err
+		}
+		for _, sct := range f.SCTData {
+			sctID, err := s.addSCTIfNotExists(tx, sct)
+			if err != nil {
+				return err
+			}
+			if err = s.addSCTFeedbackIfNotExists(tx, chainID, sctID); err != nil {
+				return err
+			}
+		}
+	}
+	return tx.Commit()
```

Comment:
Moved this (and the same pattern in the other method which used a TX) to the `defer` block.